core/src/main/java/hudson/model/Job.java

```diff
@@ -723,7 +723,9 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
 
         for (Range r : rs.getRanges()) {
             for (RunT b = getNearestBuild(r.start); b!=null && b.getNumber()<r.end; b=b.getNextBuild()) {
-                builds.add(b);
+                if (!builds.contains(b)) {
```

Comment:
contains will slow down as the size of builds increases.

The use of a `List` rather than a `Set` implies that this is not expected to be unique - even if unusual and unexpected.

I'm not sure what the typical `RangeSet of things requested here are, if the use of a `LinkList.contains()` will be an issue or not.  If it is then better to use a form of set then convert to a list before returning, Or introduce `public Set getUniqueBuilds(RangeSet)`